psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,33 +788,798 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                            for (int s = 0; s < Ssize; s++) {
+#pragma omp atomic
+                                Kp[q + Qoff][s + Soff] += K4p[(q + Qoff2) * dSsize + s + Soff2];
+                                if (!lr_symmetric_) {
+#pragma omp atomic
+                                    Kp[s + Soff][q + Qoff] += K8p[(s + Soff2) * dQsize + q + Qoff2];
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+        }  // End stripe out
+        // if (thread == 0) timer_off("JK: Atomic");
+
+    }  // End master task list
+
+    if (build_J) {
+        for (auto Jmat : J) {
+            Jmat->scale(2.0);
+            Jmat->hermitivitize();
+        }
+    }
+
+    if (build_K && lr_symmetric_) {
+        for (auto Kmat : K) {
+            Kmat->scale(2.0);
+            Kmat->hermitivitize();
+        }
+    }
+
+    if (bench_) {
+        auto mode = std::ostream::app;
+        auto printer = PsiOutStream("bench.dat", mode);
+        size_t ntri = nshell * (nshell + 1L) / 2L;
+        size_t possible_shells = ntri * (ntri + 1L) / 2L;
+        printer.Printf("Computed %20zu Shell Quartets out of %20zu, (%11.3E ratio)\n", computed_shells,
+                        possible_shells, computed_shells / (double)possible_shells);
+    }
+
+    for (int i = 0; i < computed_integrals_.size(); i++) {
+        computed_integrals_[i].clear();
+    }
+
+    timer_off("build_JK_matrices()");
+}
+
+bool linK_sort_helper(const std::tuple<int, double>& t1, const std::tuple<int, double>& t2) {
+    return std::get<1>(t1) > std::get<1>(t2);
+}
+
+void DirectJK::build_linK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, const std::vector<SharedMatrix>& D,
+                  std::vector<SharedMatrix>& J, std::vector<SharedMatrix>& K) {
+
+    timer_on("build_linK()");
+
+    // => Zeroing <= //
+
+    for (auto Jmat : J) {
+        Jmat->zero();
+    }
+
+    for (auto Kmat : K) {
+        Kmat->zero();
+    }
+
+    // => Sizing <= //
+
+    int nshell = primary_->nshell();
+    int nthread = df_ints_num_threads_;
+
+    // => Setting up computed integrals map <= //
+    computed_integrals_.resize(nthread);
+
+    // => Task Blocking <= //
+
+    std::vector<int> task_shells;
+    std::vector<int> task_starts;
+
+    // => Atomic Blocking <= //
+
+    int atomic_ind = -1;
+    for (int P = 0; P < nshell; P++) {
+        if (primary_->shell(P).ncenter() > atomic_ind) {
+            task_starts.push_back(P);
+            atomic_ind++;
+        }
+        task_shells.push_back(P);
+    }
+    task_starts.push_back(nshell);
+
+    // => End Atomic Blocking <= //
+
+    size_t ntask = task_starts.size() - 1;
+
+    std::vector<int> task_offsets;
+    task_offsets.push_back(0);
+    for (int P2 = 0; P2 < primary_->nshell(); P2++) {
+        task_offsets.push_back(task_offsets[P2] + primary_->shell(task_shells[P2]).nfunction());
+    }
+
+    size_t max_task = 0L;
```

Comment:
Your inner for loop sets the `size` of atom `task` to be the number of basis functions on atoms up to but including N for all atoms except the last, where the size is 0.

I strongly suspect the original code was bugged, and by copying, you are going to inherit the bug too. Your estimates for `max_task` are going to be much too large in the realistic case of multiple atoms, and too small in toy cases where the final atom has many shells. Please fix this bug everywhere it appears.

I recommend you improve code readability as much as possible. If the variable names were less esoteric, this bug would have been more obvious.

Reply:
I do not think this is a bug, since `task_starts` has size `ntask+1`, all atoms are considered. To test this, I ran DirectJK on a H2O molecule with O as the last atom. The code ran correctly. However, I do agree that the variables are poorly named, and I am working on refactoring the variable names, and making the code more readable.

Reply:
That resolves one of the reasons I suspect a bug, but not the second.

For each `task`, `size` will be the number of basis functions centered on atoms up to and including atom `task`. Why?

Reply:
Ah, it is to determine function offsets when we build the J and K matrices based on atom quartets. 

Example:
Building on quartet P, Q, R, S with offsets Poff, Qoff, Roff, Soff.

Jtemp[p][q] = (pq|rs) * D[Roff + r][Soff + s]
J[Poff + p][Qoff + q] += Jtemp[p][q], and vice versa for the other terms.

Reply:
I have no idea how that can be true. `size` only is declared inside the `for` loop, so the purpose is defining `max_task`. That already disagrees with what you've said. And if the goal is to define `max_task` _and the current function isn't bugged_, it's just `nbf`. This is overcomplicated and doesn't seem to be a "max_task" in any sense.

If you think this isn't a bug, then please tell me what I'm missing.

Reply:
Ah yes, it is a bug, my bad. I made a typo in my code. It should not start at 0, but `task_starts[task]`